rfcs/0151-nixos-port-alloc.md

```diff
@@ -0,0 +1,135 @@
+depend on which logic the allocator uses to generate the port number. Only asks for
+a port and get the port to be used.
+
+The port allocator will allocate ports decremented from 65535 so it's very unlikely
+that it will reach ports under 1024. Because of the ordered nature of attrsets.
+
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This is how the module would be used:
+```nix
+{ config, lib, ... }:
+lib.mkIf config.service.foo.enable {
+    networking.ports.foo-web.enable = true;
+    service.foo.port = mkDefault config.networking.ports.foo-web.port;
+}
+```
+
+And an already working implementation of the specification:
+```nix
+{ config, lib, ... }:
+
+let
+  inherit (builtins) removeAttrs;
+  inherit (lib) mkOption types submodule literalExpression mdDoc mkDefault attrNames foldl' mapAttrs mkEnableOption attrValues;
+in
+
+{
+  options.networking.ports = mkOption {
+    default = {};
+
+    example = literalExpression ''{
+      {
+        app.enable = true;
+      }
+    }'';
+
+    description = "Build time port allocations for services that are only used internally";
+
+    apply = ports: lib.pipe ports [
+      (attrNames) # gets only the names of the ports
+      (foldl' (x: y: x // {
+        "${y}" = (ports.${y}) // {
+          port = x._port;
+        };
+        _port = x._port - 1;
+      })  {_port = 65534; }) # gets the count down of the ports
+      (x: removeAttrs x ["_port"]) # removes the utility _port entity
+    ];
+
+    type = types.attrsOf (types.submodule ({ name, config, options, ... }: {
+      options = {
+        enable = mkEnableOption "Enable automatic port allocation for service ${name}";
+        port = mkOption {
+          description = "Allocated port for service ${name}";
+          type = types.nullOr types.port;
+        };
+      };
+    }));
+  };
+
+  config.environment.etc = lib.pipe config.networking.ports [
+    (attrNames)
+    (foldl' (x: y: x // {
+      "ports/${y}" = {
+        inherit (config.networking.ports.${y}) enable;
+        text = toString config.networking.ports.${y}.port;
+      };
+    }) {})
+  ];
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- If someone externally expects to use that service directly, the port which could be used
+ to access may differ like a local IP when it's not reserved by the router so it's not
+ recommended to use this module in these cases.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Keep track of which ports have been used by services and often just seeing that
+the port is already being used by some other service when the activation logs show
+that the service failed to start.
+
+Forbid usage of common utility ports like 8080, 8081, 5000, 3000 and 3333.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+How to allocate blocks of ports so something like a torrent client can use that to
+listen for p2p traffic?
```

Comment:
You could have a "hash function" that takes either a name or a port. If it's a port, it just returns the port.
Then use a [high priority](https://github.com/NixOS/rfcs/pull/151/files#r1253423732), and check that it was possible to allocate them at the desired individual port numbers, throwing an error if a collision happened anyway. (ie "required" mode mentioned before)